keps/sig-cli/kustomize-reorder-plugins.md

```diff
@@ -0,0 +1,217 @@
+
+Users may need direct control of the Resource create / update / delete ordering in cases were
+sorting by Resource Type is insufficient.
+
+### Goals
+
+- Provide the ability for users to override the order that Kustomize emits Resources
+  - Used by `kubectl apply`
+  - Used by `kubectl delete`
+  
+### Non-Goals
+
+- Ensure certain Resources are *Settled* or *Ready* before other Resources
+- Ensure dependencies between Workloads
+
+## Proposal
+
+Provide a simple mechanism allowing users to override the order that
+Resource operations are applied or deleted. Leverage the plugins feature of kustomize
+in order to let the user override the default sorting plugin.  
+
+```bash
+kustomize build <xxx> --reorder=kubectlapply | kubectl apply -f -
+kustomize build <xxx> --reorder=kubectldelete | kubect delete -f -
+```
+
+or 
+
+```bash
+kubectl apply -k <xxx>
+kubectl delete -k <xxx>
+```
+
+### kubectl apply example:
+
+The `--reorder=kubectctlapply` option indicates to kustomize to load the `kubectlapplyordertransformer.yaml transformer config` ,
+itself loading a kustomize external plugin in charge of sorting and filtering kustomize build output.
+
+An example of configuration of such a kubectapplyordertransformer.yaml:
+
+```yaml
+apiVersion: kustomize.config.k8s.io/v1beta1
+kind: KubectlApplyOrderTransformer
+metadata:
+  name: kubectlapplyordertransformer
+kindorder:
+- Namespace
+- CustomResourceDefinition
+- ServiceAccount
+- ClusterRole
+- RoleBinding
+- ClusterRoleBinding
+- ConfigMap
+- Service
+- Deployment
+- CronJob
+- ValidatingWebhookConfiguration
+- APIService
+- Job
+- Certificate
+- ClusterIssuer
+- Issuer
+```
+
+A new transformer could also be implemented to filter resources during `kubectl apply` for instance:
+
+```yaml
+apiVersion: kustomize.config.k8s.io/v1beta1
+kind: KubectlApplyFilterTransformer
+metadata:
+  name: kubectlapplyfiltertransformer
+kindorder:
+- Namespace
+- CustomResourceDefinition
+- ServiceAccount
+- ClusterRole
+- RoleBinding
+- ClusterRoleBinding
+filter:
+- Namespace
+- CustomResourceDefinition
+```
+
+Another example could be the one based on a previous KEP:
+
+```yaml
+apiVersion: kustomize.config.k8s.io/v1beta1
+kind: KubectlApplyAnnotationBasedSorter
+metadata:
+  name: kubectlapplyannotationbasedsorter
+sortOrder:
+- matchAnnotations:
+    some-annotation-name-1: some-annotation-value-1
+    some-annotation-name-a: some-label-value-a
+- matchAnnotations:
+    some-annotation-name-1: some-annotation-value-1
+- matchAnnotations:
+    some-annotation-name-2: some-annotation-value-2
+```
+
+
+### kubectl delete example:
+
+The `--reorder=kubectctldelete` option indicates to kustomize to load a `kubectlapplyordertransformer.yaml transformer config`,
+itself loading a kustomize external plugin in charge of sorting and filtering kustomize build output.
+
+An example of configuration of such a transformer.
+
+```yaml
+apiVersion: kustomize.config.k8s.io/v1beta1
+kind: KubectlDeleteOrderTransformer
```

Comment:
The `KubectlDeleteOrderTransformer` looks quite similar to `KubectlApplyOrderTransformer`. Why do you need both? Can you combine them into one?

Reply:
done